### PR TITLE
KAFKA-9615: Clean up task/producer create and close

### PR DESCRIPTION
**Pull request by vvcephei**
Consolidates task/producer management.

Now, exactly one component manages the creation and destruction of Producers,
whether they are per-thread or per-task.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:49]: Pulled out of StreamThread (could have been done a long time ago, since it was a static class anyway). I didn't embed it in TaskManager, just to keep the file size lower.

I also dropped AbstractTaskCreator, since the creation of Active and Standby tasks are only similar, not exactly the same. We weren't really using the abstraction for much except de-duplicating a few field declarations. On the con side, the abstraction made it hard to see that we were requiring several arguments for Standby task creation that were actually not ever used. The indirection also made it harder to read the task creation logic.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java:28]: We never used it as an AutoCloseable, and having it makes it hard to trace the callers.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:265]: Only the creator may close the producer, but we can go ahead and call `flush()` here.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:547]: Managed fully inside the task factory now.

You'll also notice a bunch of references to the producers are similarly gone in the following lines.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1132]: Sooo many switch statements are now gone.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:936]: We don't manage the producers anymore, so we can defer to the taskManager (who will defer to the active task creator, but that's none of the thread's business)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:63]: This is a bit on the side, but there was some wacky stuff going on in here, and afaict, the only purpose of all the nullable fields was to allow including the task id in exception messages. Do we really need to do that? If so, I'll just add the logContext to the exception message instead, since it already has the task id in it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:147]: no need to add the exception to the message, since we also pass it as the cause below.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:215]: No one who has a reference to the StreamsProducer has any business calling close, so the method is gone now.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Standby tasks don't need the mainConsumer (obvious, in retrospect).

**Review by vvcephei — COMMENTED** [review]
@guozhangwang , do you mind taking a look at this refactor?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:656]: nit: These two functions are not for testing only.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:85]: In `shutdown(final boolean clean)` we should also release task producers as well right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1132]: Nice!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:63]: Actually the `logContext.logPrefix()` should have the format of `stream-thread [%s] task [%s]` already, so all the log4j entries are good. For exception messages, we can just get the `prefix` and then encode that into the exception message.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: I'm wondering if we are introducing a latency regression without EOS here: in the old code when closing without EOS we actually do nothing, and now we would block on flushing. 

On the other hand, flushing maybe needed when we close a task to make sure all the tasks' records are acked already.

If the task is in RUNNING before shutting down, we would always commit before closing, so `flush` is already called; if the task is in RESTORING / SUSPENDED there's nothing written from this task, so a `flush` is not needed. So I think it is safe to not call `flush` after all.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java:617]: We do not created new test classes of ActiveTaskCreator / StandbyTaskCreator, but we should still have those coverage to make sure the exception thrown from producers are wrapped correctly. 

Also it seems in the new code we no long rethrow -- is that intentional. I left a comment above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: The old code we re-throw exceptions whereas here we just swallow the error. Is that intentional?

File: streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java]: `null != null`?

**Review by guozhangwang — COMMENTED** [review]
I made a pass over the PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: `closeThreadProducer()`

The method below is not public -- does this one need to be public?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: `closeProducerForTask(TaskId)` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Does this need to be public?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Does this need to be public?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Sounds reasonable to me

**Review by mjsax — COMMENTED** [review]
Nice cleanup. There is quite some overlap with https://github.com/apache/kafka/pull/8215 even if both PR address a different issue.

File: streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java
Comment (abbccdda) [reply streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java]: lol

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: I don't think we need to keep this TODO, as only after a stream 3.0 is there, we shall remove the support for task producer.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Should we throw illegal state first, since we are already in an error state?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: even better `maybeCloseThreadProducer`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:63]: Yep, that's what I was thinking. I'll go ahead and do it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Ok, I'll swap this out.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:656]: Thanks for the feedback; I didn't understand this particular comment, though.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Ah, thanks for noticing this. I meant to ask about it in here.

I thought it was strange that we would only re-throw when closing the task producers, not the thread producer. It seems like we should do the same thing in both cases, but which thing should we do?

I went with an error log in both cases, but it sounds like you wanted to throw the exception instead. Should we also rethrow for the thread producer?

File: streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java]: Oops! Auto-refactoring. I already made two passes to clean these up, looks like I missed one.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Hah, doesn't need to be there at all, actually.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: If we're going to rename the method to specify that it should do exactly "close thread producer", then this check is no longer appropriate.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:85]: Yep. Good catch.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:656]: The comment line above these two method declaration says `the following functions are for test only`, but these two functions are not.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java:617]: Good catch. Actually, there's a bunch of coverage missing from TaskManager. I'll add several more tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Yeah I think we should make them consistent: previously the closing producer is within `task#close` and if it is called via `closeDirty` we should make sure it never throws. Now since it is extracted out of the `close` call we should just rethrow for both cases.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:656]: Ah, I just found what you were talking about:
```
    // below are for testing only
```

I didn't notice that up there.

**Comment by vvcephei** [comment]
Hey @guozhangwang , @mjsax , and @abbccdda ,

I've addressed all the feedback. In particular I added about 500 lines of missing tests to TaskManagerTest. It looks like we've done a fair amount of un-unit-tested refactoring in TaskManagerTest, so I figured I'd go ahead and add the missing coverage while I was adding coverage for this refactor itself. Be forewarned, the tests are... extensive.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: 1. Let's make it a warn instead of a debug.
2. The error message can be more specific here: `Error closing task producer for task {} while handling lostAll`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Ditto here about error message

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: nit: ...WhileRebalanceInProgress

**Review by guozhangwang — APPROVED** [review]
@vvcephei I made a pass over the testing code, and only have a couple minor comments.

After addressing them please feel free to merge.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:49]: Sounds good to me

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java:247]: I was wondering what's the standard for using `assertThat` vs `assertTrue`? Do we have a convention to follow?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:27]: Are the tests added in TaskManager only trying for more coverage? @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: +1

**Review by abbccdda — APPROVED** [review]
LGTM, only have some minor comments about tests. After addressing them, feel free to merge.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java:247]: `assertThat` is nicer in general, but it doesn't really matter. In this case, IDEA offered to translate, and I was already changing a lot of assertions, so I just accepted the translation.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:27]: yep, that's right.

**Comment by vvcephei** [comment]
The last commit was trivial, and the Streams tests passed locally for me, so I'm going to go ahead and merge.

**Pull request merged**
